src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -1098,6 +1099,33 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
                 sideEffects: ImmutableArray.Create<BoundExpression>(tempAssignmentX, tempAssignmentY),
                 value: binaryExpression,
                 type: boolType);
+
+            static bool canNotBeEqualToDefaultValue(
+                [NotNullWhen(returnValue: true)] ConstantValue? constantValue)
+            {
+                // This is an explicit list so new constant values are not accidentally supported without consideration.
+                // Decimal is not in the list because it is possible to have a non-default decimal constant
+                // which is equal to the default decimal (0.0m == default(decimal)).
+                return constantValue is
+                {
+                    IsDefaultValue: false,
+                    Discriminator: ConstantValueTypeDiscriminator.Boolean
```

Comment:
>Discriminator [](http://example.com/codeflow?start=20&length=13)

It looks like we're testing the original optimization with `boolean`, `int`, `double`, and now `decimal`. We should also test the optimization for the other types below if we're not already doing so.

Reply:
Thanks. Looks like integers smaller than 32 bits are not optimized because they use Int32's equality operator, so I will remove them from the list.